Rubberduck.VBEEditor/WindowsApi/CodePaneSubclass.cs

```diff
@@ -1,61 +1,70 @@
 ﻿using System;
-using System.Windows.Forms;
+using System.Runtime.InteropServices;
 using Rubberduck.VBEditor.Events;
 using Rubberduck.VBEditor.SafeComWrappers.Abstract;
 
 namespace Rubberduck.VBEditor.WindowsApi
 {
-    public class IntelliSenseEventArgs : EventArgs
+    //Stub for code pane replacement.  :-)
+    internal class CodePaneSubclass : FocusSource, IWindowEventProvider
     {
-        public static IntelliSenseEventArgs Shown => new IntelliSenseEventArgs(true);
-        public static IntelliSenseEventArgs Hidden => new IntelliSenseEventArgs(false);
-        internal IntelliSenseEventArgs(bool visible)
-        {
-            Visible = visible;
-        }
+        public event EventHandler<EventArgs> CaptionChanged;
+        public event EventHandler<KeyPressEventArgs> KeyDown;
+        private ICodePane _codePane;
 
-        public bool Visible { get; }
-    }
-
-    public class KeyPressEventArgs
-    {
-        public KeyPressEventArgs(IntPtr hwnd, IntPtr wParam, IntPtr lParam, char character = default)
+        public ICodePane CodePane
         {
-            Hwnd = hwnd;
-            WParam = wParam;
-            LParam = lParam;
-            Character = character;
-            if (character == default(char))
-            {
-                Key = (Keys)wParam;
-            }
-            else
+            get => _codePane;
+            set
             {
-                IsCharacter = true;
+                if (HasValidCodePane)
+                {
+                    _codePane.Dispose();          
+                }
+
+                _codePane = value;
             }
         }
 
-        public bool IsCharacter { get; }
-        public IntPtr Hwnd { get; }
-        public IntPtr WParam { get; }
-        public IntPtr LParam { get; }
-
-        public bool Handled { get; set; }
+        public bool HasValidCodePane
+        {
+            get
+            {
+                if (_codePane == null)
+                {
+                    return false;
+                }
 
-        public char Character { get; }
-        public Keys Key { get; }
-    }
+                try
+                {
+                    if (Marshal.GetIUnknownForObject(_codePane.Target) != IntPtr.Zero)
+                    {
+                        return true;
+                    }
 
-    //Stub for code pane replacement.  :-)
-    internal class CodePaneSubclass : FocusSource
-    {
-        public ICodePane CodePane { get; }
+                    _codePane.Dispose();
+                    _codePane = null;
+                }
+                catch
+                {
+                    // All paths leading to here mean that we need to ditch the held reference, and there
+                    // isn't jack all that we can do about it.
+                    _codePane = null;
```

Comment:
Actually we should stick in a `Debug.Warn` in here so that it'll show up in the log. Make the 🐦 go 🎤.

Reply:
Done.